include/deal.II/non_matching/fe_values.h

```diff
@@ -368,6 +369,377 @@ namespace NonMatching
+     */
+    const SmartPointer<const MeshClassifier<dim>> mesh_classifier;
+
+    /**
+     * For each element in the FECollection passed to the constructor,
+     * this object contains an dealii::FEInterfaceValues object created with a
+     * quadrature rule over the full reference cell: $[0, 1]^{dim-1}$ and
+     * UpdateFlags for the inside region. Thus, these optionals should always
+     * contain a value.
+     *
+     * When LocationToLevelSet of the cell is INSIDE (and we do not need
+     * to generate an immersed quadrature), we return the
+     * dealii::FEInterfaceValues object in this container corresponding to the
+     * cell's active_fe_index.
+     *
+     * This container is a std::deque, which is compatible with the
+     * `FEInterfaceValues` class that does not have a copy-constructor.
+     */
+    std::deque<std_cxx17::optional<dealii::FEInterfaceValues<dim>>>
+      fe_values_inside_full_quadrature;
+
+    /**
+     * For each element in the FECollection passed to the constructor,
+     * this object contains an dealii::FEInterfaceValues object created with a
+     * quadrature rule over the full reference cell: $[0, 1]^{dim-1}$ and
+     * UpdateFlags for the outside region. Thus, these optionals should always
+     * contain a value.
+     *
+     * When LocationToLevelSet of the cell is OUTSIDE (and we do not need
+     * to generate an immersed quadrature), we return the dealii::FEValues
+     * object in this container corresponding to the cell's active_fe_index.
+     *
+     * This container is a std::deque, which is compatible with the
+     * `FEInterfaceValues` class that does not have a copy-constructor.
+     */
+    std::deque<std_cxx17::optional<dealii::FEInterfaceValues<dim>>>
+      fe_values_outside_full_quadrature;
+
+    /**
+     * FEInterfaceValues object created with a quadrature rule integrating over
+     * the inside region, $\{x \in B : \psi(x) < 0 \}$, that was generated in
+     * the last call to reinit(..). If the cell in the last call was not
+     * intersected or if 0 quadrature points were generated, this optional will
+     * not contain a value.
+     */
+    std_cxx17::optional<dealii::FEInterfaceValues<dim>> fe_values_inside;
+
+    /**
+     * FEInterfaceValues object created with a quadrature rule integrating over
+     * the outside region, $\{x \in B : \psi(x) > 0 \}$, that was generated in
+     * the last call to reinit(..). If the cell in the last call was not
+     * intersected or if 0 quadrature points were generated, this optional will
+     * not contain a value.
+     */
+    std_cxx17::optional<dealii::FEInterfaceValues<dim>> fe_values_outside;
+
+    /**
+     * Object that generates the immersed quadrature rules.
+     */
+    DiscreteFaceQuadratureGenerator<dim> quadrature_generator;
+  };
+
+
+#ifndef DOXYGEN
+
+  /*---------------------- Inline functions ---------------------*/
+
+  template <int dim>
+  template <class CellIteratorType>
+  void
```

Comment:
```suggestion
  inline void
```

same below

Reply:
Why are the implementations of these functions in the header file?

Reply:
I don't know how one would instantiate all the cell-types to also work with the iterator-filters. The corresponding function in `dealii::FEInterfaceValues` are in the header too. See #13169 and #13170.

Reply:
Than let's keep it here. But the inline keyword should probably nevertheless there!